patches/api/0327-Add-PlayerCompostItemEvent.patch

```diff
@@ -0,0 +1,112 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Noah van der Aa <ndvdaa@gmail.com>
+Date: Sat, 7 Aug 2021 15:11:27 +0200
+Subject: [PATCH] Add PlayerCompostItemEvent.
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerCompostItemEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerCompostItemEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1195392aff2829eed6d987d1896b3fe453cafc53
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerCompostItemEvent.java
+@@ -0,0 +1,100 @@
++package com.destroystokyo.paper.event.player;
++
++import org.bukkit.block.Block;
++import org.bukkit.entity.Player;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by a player.
++ */
++public class PlayerCompostItemEvent extends PlayerEvent implements Cancellable {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++
++    private Block composter;
++    private ItemStack item;
++    private boolean willRaiseLevel;
++    private int level;
++
++    private boolean cancelled;
++
++    public PlayerCompostItemEvent(@NotNull Player who, @NotNull Block composter, @NotNull ItemStack item, boolean willRaiseLevel, int level) {
++        super(who);
++        this.composter = composter;
++        this.item = item;
++        this.willRaiseLevel = willRaiseLevel;
++        this.level = level;
++    }
++
++    /**
++     * Gets the composter that was interacted with.
++     *
++     * @return the composter
++     */
++    @NotNull
++    public Block getComposter() {
++        return this.composter;
++    }
++
++    /**
++     * Gets the item that was used on the composter.
++     *
++     * @return the item
++     */
++    @NotNull
++    public ItemStack getItem() {
++        return this.item;
++    }
++
++    /**
++     * Gets whether or not the composter will rise a level.
++     *
++     * @return true if successful
++     */
++    public boolean willRaiseLevel() {
++        return this.willRaiseLevel;
++    }
++
++    /**
++     * Sets whether or not the composter will rise a level.
++     *
++     * @param willRaiseLevel true if the composter should rise a level
++     */
++    public void setWillRaiseLevel(boolean willRaiseLevel) {
++        this.willRaiseLevel = willRaiseLevel;
++    }
++
++    /**
++     * Gets the current composter level.
++     *
++     * @return the composter level
++     */
++    public int getLevel() {
```

Comment:
I feel like this is dangerous and could lead to de-sync (from the plugin author's perspective) between this value and the block's actual BlockData. What if an earlier event modifies the actual level of the composter?

Reply:
Good catch, I'll remove this since the block can be casted to [Levelled](https://papermc.io/javadocs/paper/1.17/org/bukkit/block/data/Levelled.html) to get the level.

Reply:
Yeah, it’s hard to say what to do actually, because if you set the level that way, and don’t cancel the event, it will have practically no effect because it’s set after the event as well. Idk the best course of action here tbh.

Reply:
I've removed `getLevel()` for now, not sure if it should be added back.

Reply:
You know, perhaps a getNewLevel()/setNewLevel() pair of methods would be useful. I know right now its only going to increase by one, but make a note in the javadoc for such methods that changing the level on the BlockData will be overriden unless the event is cancelled.